openedx/core/djangoapps/oauth_dispatch/docs/decisions/0008-use-asymmetric-jwts.rst

```diff
@@ -0,0 +1,186 @@
+capability cannot be achieved unless verifiable tokens can be forwarded from one service to another. 
+
+Decisions
+---------
+
+Asymmetric JWTs
+~~~~~~~~~~~~~~~
+
+We will introduce identified "asymmetric" cryptographic keys for signing JWTs. The OAuth Provider will
+be the only service configured with the aymmetric keypair, including its Private and Public key portions.
+All other OAuth Clients will be configured with only the Public key portion of the asymmetric key pair.
+
+"kid" Key Identifier
+~~~~~~~~~~~~~~~~~~~~
+
+In order to support key rotation in a forward compatible manner, we will identify the asymmetric keys,
+using the `JSON Web Key (JWK)`_ standard's `"kid" (Key ID)`_ parameter.  When a `JSON Web Signature (JWS)`_
+is created to sign a JWT, its `"kid" header parameter`_ specifies which key was used to secure the JWS.
+
+.. _JSON Web Key (JWK): https://tools.ietf.org/html/draft-ietf-jose-json-web-key-36
+.. _`"kid" (Key ID)`: https://tools.ietf.org/html/draft-ietf-jose-json-web-key-36#section-4.5
+.. _JSON Web Signature (JWS): https://tools.ietf.org/html/rfc7515
+.. _`"kid" header parameter`: https://tools.ietf.org/html/rfc7515#section-4.1.4
+
+Remove JWT_ISSUERS
+~~~~~~~~~~~~~~~~~~
+
+edx_rest_framework_extensions.settings_ supports having a list of **JWT_ISSUERS** instead of just a single
+one. This support for configuring multiple issuers is present across many services. However, this does not
+conform to the `JWT standard`_, where the `issuer`_ is intended to identify the entity that generates and
+signs the JWT. In our case, that should be the single Auth service only.
+
+If different values for the issuer_ claim are needed for multi-tenancy purposes, those should be specified
+using `site configuration`_ variants instead of adding complexity with multiple issuers.
+
+Additionally, **JWT_ISSUERS** is not intended to be used for key rotation. Rather, the set of active signing
+keys should be specified as a `JSON Web Key Set (JWK Set)`_ instead. Thus, there would only be a single
+issuer, but with (the potential of) multiple signing keys stored in a JWT Set.
+
+.. _edx_rest_framework_extensions.settings: https://github.com/edx/edx-drf-extensions/blob/1db9f5e3e5130a1e0f43af2035489b3ed916d245/edx_rest_framework_extensions/settings.py#L73
+.. _JWT standard: https://tools.ietf.org/html/rfc7519
+.. _issuer: https://tools.ietf.org/html/rfc7519#section-4.1.1
+.. _JSON Web Key Set (JWK Set): https://tools.ietf.org/html/draft-ietf-jose-json-web-key-36#section-5
+.. _site configuration: https://github.com/edx/edx-platform/blob/af841336c7e39d634c238cd8a11c5a3a661aa9e2/openedx/core/djangoapps/site_configuration/__init__.py
+
+Example Code
+------------
+
+KeyPair Generation
+~~~~~~~~~~~~~~~~~~
+
+Here is code for generating a keypair::
+
+    from Cryptodome.PublicKey import RSA
+    from jwkest import jwk
+
+    rsa_key = RSA.generate(2048)
```

Comment:
Do we know that this defaults to a SHA2 signed cert/key?  (The old default, even when you said 2048 was often SHA1).  I know openssl's defaults, but not the python defaults.

Reply:
Ah, got it, we're just generating a keypair, not anything that's verified to be part of a chain.
I saw the signatures below which I was curious how this differed.